Netdev2

scottfeldman opened the pull request:
This PR adds a network device driver model called netdev.  There will be a companion PR for TinyGo drivers to update the netdev drivers and network examples.  This PR covers the core "net" package.

An RFC for the work is here: #https://github.com/tinygo-org/drivers/issues/487.  Some things have changed from the RFC, but nothing major.

The "net" package is a partial port of Go's "net" package, version 1.19.3.  The src/net/README file has details on what is modified from Go's "net" package.

Most "net" features are working as they would in normal Go.  TCP/UDP/TLS protocol support is there.  As well as HTTP client and server support.  Standard Go network packages such as golang.org/x/net/websockets and Paho MQTT client work as-is.  Other packages are likely to work as-is.

Testing results are [here](https://docs.google.com/spreadsheets/d/e/2PACX-1vT0cCjBvwXf9HJf6aJV2Sw198F2ief02gmbMV0sQocKT4y4RpfKv3dh6Jyew8lQW64FouZ8GwA2yjxI/pubhtml?gid=1013173032&single=true).

aykevl committed:
runtime: check for heap allocations inside interrupts

This is unsafe and should never be done.

aykevl committed:
ci: build LLVM with thread support on Windows

This should fix a number of concurrency/threading issues.

I had to force-disable concurrency in the linker using a hack. I'm not
entirely sure what the cause is, possibly the MinGW version (version 12
appears to work for me, while version 11 as used on the GitHub runner
image seems to be broken).
There are a few ways to fix this in a better way:
  * Fix the underlying cause (possibly by upgrading to MinGW-w64 12).
  * Add the `--threads` flag to the LLD MinGW linker, so we can use a
    regular parameter instead of this hack.

aykevl committed:
all: refactor reflect package

This is a big commit that changes the way runtime type information is stored in
the binary. Instead of compressing it and storing it in a number of sidetables,
it is stored similar to how the Go compiler toolchain stores it (but still more
compactly).

This has a number of advantages:

  * It is much easier to add new features to reflect support. They can simply
    be added to these structs without requiring massive changes (especially in
    the reflect lowering pass).
  * It removes the reflect lowering pass, which was a large amount of hard to
    understand and debug code.
  * The reflect lowering pass also required merging all LLVM IR into one
    module, which is terrible for performance especially when compiling large
    amounts of code. See issue 2870 for details.
  * It is (probably!) easier to reason about for the compiler.

The downside is that it increases code size a bit, especially when reflect is
involved. I hope to fix some of that in later patches.

aykevl committed:
builder: add test to check for changes in binary size

This test only applies when using the built-in LLVM version. This way,
we have a stable LLVM version to test against. Distribution versions of
LLVM (especially Debian) tend to be patched in a way that affect the
results.

aykevl committed:
all: use unsafe.Add instead of unsafe.Pointer(uintptr(...) + ...)

We have an optimization for this specific pattern, but it's really just
a hack. With the addition of unsafe.Add in Go 1.17 we can directly
specify the intent instead and eventually remove this special case.

The code is also easier to read.

aykevl committed:
reflect: make sure null bytes are supported in tags

deadprogram committed:
machine/rp2040: implement Flash interface

Signed-off-by: deadprogram <ron@hybridgroup.com>

kenbell committed:
arm: enable functions in RAM for go & cgo

ysoldak committed:
rp2040: rtc delayed interrupt

waj334 committed:
Added missing TCPAddr and UDPAddr implementations to the net package

scottfeldman committed:
fresh netdev checkin preparing for PR

scottfeldman committed:
fresh netdev checkin preparing for PR

scottfeldman committed:
Documentation and cleanup of net package

scottfeldman committed:
Documentation and cleanup of net package

scottfeldman committed:
fix goroot override paths to pickup all of net package

scottfeldman committed:
fix goroot override paths to pickup all of net package

scottfeldman committed:
gofmt -l -w -s

scottfeldman committed:
gofmt -l -w -s

dgryski committed:
compiler, reflect: add map key and element type info

scottfeldman committed:
cleanup to sync with 1.19.3 Go as much as possible

scottfeldman committed:
cleanup to sync with 1.19.3 Go as much as possible

bgould commented on src/net/net.go:
```diff
@@ -7,10 +7,19 @@
-	"io"
 	"time"
+
+	"tinygo.org/x/drivers/netdev"
```
I think maybe this shouldn't be referencing the drivers module; it might be preferable to have a copy of the netdev.Netdever interface in TinyGo instead of in `drivers` (or alternatively, have the primary definition be in TinyGo and referenced from `drivers`).

scottfeldman replied:
I had it that way at one point, but changed it as to not add any new exports to "net" package.

deadprogram replied:
I understand the motivation, but we for sure do not wish for the circular reference. Either use a subpackage, just add another file with the exported interface, or copy the interface definition into both the tinygo repo and also a duplicate the drivers package.

scottfeldman replied:
Ok, new commit moving drivers/netdev to tinygo/src/net/netdev.  This should decouple "net" and "netdev" and remove the circular reference.

soypat replied:
If we do want to have netdev package inside of `Tinygo` it should be unreachable by external users. You can do this by putting `netdev` directory under another directory called `internal`. See [Dave Cheney's](https://dave.cheney.net/2019/10/06/use-internal-packages-to-reduce-your-public-api-surface) blog post.

bgould commented on src/net/udpsock.go:
```diff
@@ -0,0 +1,249 @@
+// UDPConn is the implementation of the Conn and PacketConn interfaces
+// for UDP network connections.
+type UDPConn struct {
+	fd            netdev.Sockfd
```
Similar comment as for netdev.Netdever ... Sockfd should maybe be defined in the TinyGo repo instead of `drivers`.  Also wondering if Sockfd should be `uintptr` instead of `int` (in `os` for instance file descriptors are `uintptr`)

scottfeldman replied:
int was what socket(2) calls return, so just copied that.  And -1 is a common error return value for socket(2).

bgould commented:
Great work, this is very exciting :)

deadprogram commented:
Seems like some dependency problem here @scottfeldman https://github.com/tinygo-org/tinygo/actions/runs/4190521808/jobs/7264005556#step:13:32

deadprogram committed:
machine/rp2040: change calling order for device enumeration fix to do first

Signed-off-by: deadprogram <ron@hybridgroup.com>

anuraaga committed:
runtime: allow custom-gc SetFinalizer and clarify KeepAlive

scottfeldman commented:
@deadprogram I'm stuck.  Both netdev PRs for tinygo and drivers builds are failing because they depend on each other.  tinygo needs to import drivers/netdev and drivers needs to import tinygo "net" for net.useNetdev.  These build fine in my go workspace where they're both present.  I'm at a loss on what to do.

aykevl committed:
runtime: implement KeepAlive using inline assembly

scottfeldman commented:
@deadprogram I started over in a fresh go workspace with a fresh clone of tinygo and drivers,  checked out netdev branch for each and built everything again.  Ran make test on tinygo and drivers and both PASS.

It seems both PRs need to be merged at the same time and folks would have to refresh both to avoid hitting the build issue we're hitting here.

deadprogram committed:
build: always cache LLVM source/build even if the tests fail to avoid extra rebuilds (#3453)

builds/macos, linux, windows: update to explicit restore/save for LLVM source and LLVM builds

scottfeldman committed:
add missing pipe.go back; it got left out in netdev conversion

scottfeldman committed:
add missing pipe.go back; it got left out in netdev conversion

scottfeldman committed:
add some net package tests for smoketest

scottfeldman committed:
add some net package tests for smoketest

sago35 committed:
machine/usb/hid: add MediaKey support (#3436)

machine/usb/hid: add MediaKey support

andyshinn committed:
fix bad qt py pin assignment

ya-mouse committed:
Better handling of sub-clusters in SVD (#3335)

gen: Better handling of sub-clusters in SVD

deadprogram committed:
build/linux: install wasmtime directly from github release

Signed-off-by: deadprogram <ron@hybridgroup.com>

deadprogram committed:
build/docker: trigger all builds in tinygo ecosystem repos using GH actions

Signed-off-by: deadprogram <ron@hybridgroup.com>

deadprogram committed:
build/docker: use build context from pre-saved LLVM container to avoid rebuilding

Signed-off-by: deadprogram <ron@hybridgroup.com>

deadprogram committed:
docker: ignore changes to CI itself to avoid breaking cache on images

Signed-off-by: deadprogram <ron@hybridgroup.com>

aykevl committed:
main: support qemu-user debugging

Running binaries in QEMU (when debugging on Linux for example) did not
work correctly as qemu-user expects the `-g` flag to be first on the
command line before the program name. Putting it after will make it a
command line parameter for the emulated program, which is not what we
want.

I don't think this ever worked correctly.

aykevl committed:
arm64: fix register save/restore to include vector registers

Some vector registers must be preserved across calls, but this wasn't
happening on Linux and MacOS. When I added support for windows/arm64, I
saw that it required these vector registers to be preserved and assumed
this was Windows deviating from the standard calling convention. But
actually, Windows was just implementing the standard calling convention
and the bug was on Linux and MacOS.

This commit fixes the bug on Linux and MacOS and at the same time merges
the Go and assembly files as they no longer need to be separate.

aykevl committed:
ci: don't pass -v to `go test`

It can be difficult to find what went wrong in a test. Omitting -v
should make it easier to see the failing tests and the output for them
(note that output is still printed for tests that fail).

inindev committed:
Pins D4 & D5 are I2C1. Use pins D2 & D3 for I2C0.

Signed-off-by: John Clark <inindev@gmail.com>

deadprogram committed:
machine/lorae5: add needed definition for UART2

Signed-off-by: deadprogram <ron@hybridgroup.com>

scottfeldman committed:
move netdev from drivers

scottfeldman committed:
move netdev from drivers

scottfeldman committed:
README: consolidate net and netdev readmes

scottfeldman committed:
README: consolidate net and netdev readmes

scottfeldman committed:
remove parse_test.go because on file support

scottfeldman committed:
remove parse_test.go because on file support

scottfeldman committed:
remove parse_test.go because on file support

scottfeldman committed:
remove parse_test.go because on file support

aykevl committed:
internal/task: disallow blocking inside an interrupt

Blocking inside an interrupt is always unsafe and will lead to all kinds
of bad behavior (even if it might appear to work sometimes). So disallow
it, just like allocating heap memory inside an interrupt is not allowed.

I suspect this will usually be caused by channel sends, like this:

    ch <- someValue

The easy workaround is to make it a non-blocking send instead:

    select {
    case ch <- someValue:
    default:
    }

This does mean the application might become a bit more complex to be
able to deal with this case, but the alternative (undefined behavior) is
IMHO much worse.

Commit:
Rearrange switch case for get pin cfg

soypat committed:
rp2040: provide better errors for invalid pins on i2c and spi (#3443)

machine/rp2040: provide better errors for invalid pins on i2c and spi

deadprogram committed:
machine/rp2040: correct issue with i2c pin validation

Signed-off-by: deadprogram <ron@hybridgroup.com>

deadprogram committed:
machine/rp2040: correct issue with spi pin validation

Signed-off-by: deadprogram <ron@hybridgroup.com>

soypat commented on src/net/netdev/socket.go:
```diff
@@ -0,0 +1,143 @@
+}
+
+func NewSockAddr(host string, port Port, ip IP) SockAddr {
+	return SockAddr{host: host, port: port, ip: ip}
```
A lot of these types are shared by the Go standard library. Maybe it's best to reuse them? `net.IP` and friends I mean. `net.IP` is a slice of bytes, If we keep this whole package outside of `net` we can simply use the existing `net` package for all these types. Port is also represented as an `int` in `net` (see `TCPAddr`).

scottfeldman replied:
Doing so would create a circular dependency...

"net" imports "netdev"
"netdev" imports "net"

One way to work around this is to put everything in the "net" package.   So netdev.go and socket.go would just be part of the "net" package; there would be no netdev package.  I'm not a fan of this solution as it adds all those netdev exports to the net package.  The current PR has just the one export added to net: net.UseNetdev() and netdev package does not import "net".

Keeping netdev not dependent on net makes it easier to move its location around (current at net/netdev), and let's one use netdev without net package, if desired.  See rawsocket example in this PR.  It doesn't use "net" at all.

soypat replied:
Snap, you're right. Then probably better just to use a `[]byte` slice for IP and an int for port. We should still go back to how we were doing it when we got the wifinina working. Can we go that way without causing circular dependency? That and also avoid exporting `UseNetdev` from `net`'s side.

soypat commented on src/net/netdev/netdev.go:
```diff
@@ -0,0 +1,133 @@
+	"strings"
+)
+
+type HardwareAddr []byte
```
`net` already has a `HardwareAddr`! Let's use it! I understand you are not using it here to avoid a cyclic dependency- well that's another reason to keep `netdev` in its own external package for now, either in `drivers` or `tinyio` when https://github.com/tinygo-org/drivers/issues/491 is approved, if it is approved.

scottfeldman replied:
Same circular dependency issue.

soypat commented on src/net/net.go:
```diff
@@ -7,10 +7,18 @@
+var dev netdev.Netdever
+
+// UseNetdev sets the netdev
+func UseNetdev(n netdev.Netdever) {
```
Again, this should not be exported. Link at compile time! Our objective should be to maintain compatibility during the rollout of this shiny brand new feature. OVer time we will learn how it works and what we can do to make it better, maybe even adding in this function as exported. But best to play it safe for now. Unexporting it later may break programs.

scottfeldman replied:
Hmmm...let me check if the //go:linkname trick you introduced would work when the netdev package lives in tinygo.  That would hide the UseNetdev() export from net package.

Let me play around with it and see...

soypat commented on src/net/tlssock.go:
```diff
@@ -0,0 +1,156 @@
+// A TLSConn represents a secured connection.
+// It implements the net.Conn interface.
+type TLSConn struct {
+	fd            netdev.Sockfd
```
If a Sockfd is just a plain old integer lets use a `int32` or similar for now and call all instances of it `fd*` in code.

soypat commented on src/net/udpsock.go:
```diff
@@ -0,0 +1,248 @@
+		return nil, err
+	}
+
+	var ip netdev.IP
```
Can this be just a plain `net.IP`?

soypat commented on src/net/netdev/socket.go:
```diff
@@ -0,0 +1,143 @@
+// Berkely Sockets-like interface.  See man page for socket(2), etc.
+//
+// Multiple goroutines may invoke methods on a Socketer simultaneously.
+type Socketer interface {
+	Socket(family AddressFamily, sockType SockType, protocol Protocol) (Sockfd, error)
+	Bind(sockfd Sockfd, myaddr SockAddr) error
+	Connect(sockfd Sockfd, servaddr SockAddr) error
+	Listen(sockfd Sockfd, backlog int) error
+	Accept(sockfd Sockfd, peer SockAddr) (Sockfd, error)
+	Send(sockfd Sockfd, buf []byte, flags SockFlags, timeout time.Duration) (int, error)
+	Recv(sockfd Sockfd, buf []byte, flags SockFlags, timeout time.Duration) (int, error)
+	Close(sockfd Sockfd) error
+	SetSockOpt(sockfd Sockfd, level SockOptLevel, opt SockOpt, value any) error
+}
```
## On the `Socketer`'s method types
1. Keep as is with own types.

2. Use `net` types where possible or primitive types everywhere else. I like this one because its simple. Berkeley sockets are well known. You can define the Constants in the `net` package for the SockOptLevel, SockFlag, Protocol, etc. 
  ```go
  type Sockfd = int32
  type SockFlags = int
  type SockOptLevel = int
  type SockOpt = int
  // Decompose SockAddr
  type IP = net.IP 
  type Port = int    // like net package
  ```
  You can interchange some `int`s for `uint`s if you know they are only positive to give even better type hints to users.

3. Find a middle ground. I've seen that some types have `String` methods. That is fantastic for debuggability and developer happiness. We should probably keep that. That said- `SockAddr` should definetely be a `ip net.IP` and a `port int`, just like the standard library. `Sockfd` looks like it should be a `int32`, there is not much to be gained from giving it a type. It's also pretty apparent what is and isn't a FD from looking at bare source code (usually called `fd` or `fd*`). If we can't add meaningful methods to SockOptLevel, SockFlags, SockOpt, then they or some of them should be integers of fixed size according the the berkeley API.

scottfeldman replied:
option 2 is out because of circular dep.

Sockfd is opaque to client code; client code shouldn't understand what's inside or how big it is.  It's passed as an opaque token.

I find user-defined types self-documenting and I like the compiler checks I gain with user-defined types, so I'm having a hard time backing away from that.  Is there a performance (size, time) issue using user-defined types?  I don't believe there is any penalty in the hot path (send/recv).

soypat commented on src/net/netdev/netdev.go:
```diff
@@ -0,0 +1,133 @@
+// interfaces to the netdev driver directly to provide TCPConn, UDPConn,
+// and TLSConn socket connections.
+
+package netdev
```
I am not in favor regarding the addition of a package that would break native-tinygo compatibility.

soypat replied:
As I mentioned above, we can put `netdev` under an internal directory as [per this link](https://dave.cheney.net/2019/10/06/use-internal-packages-to-reduce-your-public-api-surface).

soypat reviewed (changes requested):
I really appreciate the work you've put into this whole PR. This is exactly where I hoped TinyGo would be in the future a while ago and now it's here- it's unreal. Now being so close to the finish line I really want to take the time to really sit down and make sure we take TinyGo in the right direction- and that may take some thinking.

Right now I'm thinking this new package is not quite what I envisioned for TinyGo. I don't think the netdev package should live inside the standard library unless native Go incorporated it. IT would fundamentally split the ecosystem. We should strive to keep Gophers and Tinygophers united if possible.

Maybe adding this package is the only way of implementing Netdev... but I think it isn't. In fact, I'm quite sure we can do a better job at keeping compatibility with upstream/native. More comments to follow.

Lastly I want to link an issue that is related to this work https://github.com/tinygo-org/drivers/issues/491. This RFC is for the creation of a `tinyio` package that would solve many of the dependency cycle issues we are having, including in the `machine` package. I believe `netdev` would be a perfect fit as a subpackage of `tinyio` since it is neither a driver nor a stdlib package, but rather a bridge between the two. i.e. an interface, which is what `tinyio` seeks to standardize.

dgryski committed:
runtime: properly turn pointer into empty interface when hashing

aykevl committed:
ci: add AVR timers test

Add the timers test because they now work correctly on AVR, probably as
a result of the reflect refactor: https://github.com/tinygo-org/tinygo/pull/2640

I've also updated a few of the other tests to indicate the new status
and why they don't work. It's no longer because of compiler errors, but
because of linker or runtime errors (which is at least some progress).
For example, I found that testdata/reflect.go works if you disable
`testAppendSlice` and increase the stack size.

aykevl commented on src/net/README.md:
```diff
@@ -0,0 +1,473 @@
+
+## Using "net/http" Package
+
+TinyGo's net/http package is a partial port of Go's net/http package, providing
+a subset of the full net/http package.  There are a few features excluded
+during the porting process, in particular:
+
+- No HTTP/2 support
+- No TLS support for HTTP servers (no https servers)
+- HTTP client request can't be reused
```
Can you explain why replacing net/http is necessary?

scottfeldman replied:
> I agree we really need to have networking support, but I'm worried the code in this PR is going to be very difficult to maintain. Currently we need a few changes here and there to support a new Go version (especially when that Go version adds new language features), but so far we managed to avoid having to pull code from upstream on each release. This PR would change that in a large way.

There is a maintenance burden, but I didn't see it as difficult.  There are 10 files under src/net, 5 are straight copies, the other 5 are modified.  Someone needs periodically to back-port upstream changes for bug/security fixes.  As part of this PR, I would offer to maintain these files.  Maybe it's just because I've been thru these 10 files so much, it seems easy to maintain.

> In particular, I wonder whether we could keep using the Go stdlib version of packages like net/http and crypto/tls, and if that's not appropriate for microcontrollers maybe have a separate package with lower resources (outside the standard library)? This is especially important for net/http:

tls/crypto: no.  Go's tls/crypto is software support for TLS protocol.  We have mbed TLS offload where the co-controller (e.g. ESP32) handles the TLS protocol.  netdev goes straight to the offload model.

net/http: good luck, I guess is my first thought.  I've tried multiple times to get the full net/http code working without modifications.  Each time I abandon the effort because of too many things are missing in TinyGo.  I should have keep better notes so I could give you details.  I wouldn't mind revisiting if we can make some progress on this PR.  Other packages such as net/url, net/textproto, net/rpc, net/mail, net/smtp just work, as well as the golang.org/x/net (websocket) packages and third party packages (Paho MQTT, etc).

> 1. I expect there are many people who would like to have a fully featured net/http package (with TLS and all that).

There is TLS for net/http, but only on the client side.  So client http.Get/Head/Post/PostForm() calls work for http or https (TLS).  Http server TLS support is not there.  So no http.ListenAndSeverTLS(), but http.ListenAndServe() is OK.

It should be possible to add https server support, if someone needed it.  The mbed TLS offload on these co-controllers we're using can support it.  I'm inclined to wait and see if someone actually needs that support.  I guess it would be cool to run an https on a micro-controller with let's encrypt and plug it directly into the internet?  Are people running internet services on micro-controllers?   My thinking was these embedded devices live on the edge of the internet, not on the interior.  But there is a path forward to move from the edge.

> 2. If we copy+modify upstream net/http that will also put a burden on us to fix any security issues that might arise. I'm not sure we're able to do that.

There are 17 files in net/http, 9 of which are straight copies, and the other 8 are modified.  I can make same offer to maintain these files wrt security and other upstream changes.

bgould replied:
Sort of talking out of my assumptions here ... so feel free to shoot it down if this is way off base  ... but I wonder it _might_ be feasible use the standard `net` package by linking implementations of syscalls that dispatch to `netdev` implementations?

As an example of what I mean ... NestedVM was a compiler for producing Java binaries from C/C++ code ... in that case, POSIX functionality was integrated using [newlib](https://sourceware.org/newlib/) ... newlib allows users to implement a callback for dispatching syscalls including network ones such as: https://github.com/bgould/nestedvm/blob/master/src/org/ibex/nestedvm/UnixRuntime.java#L174-L185

Not sure if this is feasible ... but if so, I would imagine that the relevant syscall definitions in standard Go should be relatively stable ... almost certainly more stable that the `net` package.

soypat replied:
While I do see value in merging with upstream Go that may be a herculean task we're asking for. I think we should strive to get this as polished as possible and completely Go-compatible as possible so that in the future we can merge TinyGo net package with upstream net package with not so much as a hiccup. This would mean eliminating any exported function signatures that are not in upstream Go and keeping the Application usage the same, which I believe is identical, if not very similar.

scottfeldman replied:
> Sort of talking out of my assumptions here ... so feel free to shoot it down if this is way off base ... but I wonder it _might_ be feasible use the standard `net` package by linking implementations of syscalls that dispatch to `netdev` implementations?

This would be the ultimate solution.  Replacing syscall/sockets with netdev/sockets at the portability layer in the net package.  I tried that route and didn't get very far.  I saved the branch in case I wanted to revisit, but if I recall I was bumping into things that were not available in tinygo.  Even the crypto/tls package should port, negating the need to offload TLS, assuming we can do all the crypto required on these controllers.  Let's say all of this is possible, and someone can pull this off in the future.  What would be the migration path from this PR to the ultimate solution?  Let's look at the pieces:

--------

#### Apps
In this PR, apps import "net", "net/http" and "crypto/tls", same as they would if running in normal Go env.  Nothing should change from the apps' perspective when migrating to the ultimate solution.  Users are unaware the rug has been pulled out, upgraded, and slipped back in.

#### Drivers
In this PR, drivers implement the netdev interface, which is basically the BSD socket(2) interface.  The ultimate solution ports the "net" package to netdev rather than syscall/sockets, and there is a good chance nothing changes in the netdev interface during this port, so nothing changes in the drivers.  (BSD sockets will probably be with us forever).  We retain our investment in drivers.

#### Interfaces
There is a loose end with net.UseNetdev() where the app must pick which netdev driver to use.  We don't really have support for network interfaces (eth0, wlan0, etc); we're hard wiring the driver to the app with UseNetdev().  If the ultimate solution can mimic the full OS network interfaces as defined in the "net" package, then maybe UseNetdev moves outside of the apps visibility.  So there would be a change required to the app.  But to support a full OS model of interfaces, there would need to be something equivalent to the "ip" command in Linux to configure and manage the interfaces.  Seems way overkill for embedded.

--------

So I'm feeling pretty good about a migration path forward from this PR.  I'm skeptical the full "net" package could be ported at the syscall level, but just in case it can, I think we have a good migration story.  

If someone wants to take a go at a full "net" package port, this PR would be a good starting point as the netdev driver parts are done.  In my attempt, I started with an empty tinygo/src/net directory and tried to compile one of the drivers/examples/net examples.  Adding/modifying "net" package files as needed to get a clean compile.  Adding xxx_tinygo.go files where the xxx_[posix|linux|unix|windows].go files reside, and so on.

aykevl reviewed (commented):
I agree we really need to have networking support, but I'm worried the code in this PR is going to be very difficult to maintain. Currently we need a few changes here and there to support a new Go version (especially when that Go version adds new language features), but so far we managed to avoid having to pull code from upstream on each release. This PR would change that in a large way.

In particular, I wonder whether we could keep using the Go stdlib version of packages like net/http and crypto/tls, and if that's not appropriate for microcontrollers maybe have a separate package with lower resources (outside the standard library)?
This is especially important for net/http:
1. I expect there are many people who would like to have a fully featured net/http package (with TLS and all that).
2. If we copy+modify upstream net/http that will also put a burden on us to fix any security issues that might arise. I'm not sure we're able to do that.

dgryski committed:
reflect: add Type.Elem() and Type.Key() for Maps

dgryski committed:
compiler: zero struct padding during map operations

Fixes #3358

dgryski committed:
runtime: add unsafe.pointer reflect wrappers for hashmap calls

dgryski committed:
reflect: add MapIndex()

dgryski committed:
reflect: add MapRange/MapIter

scottfeldman commented:
> Right now I'm thinking this new package is not quite what I envisioned for TinyGo. I don't think the netdev package should live inside the standard library unless native Go incorporated it. IT would fundamentally split the ecosystem. We should strive to keep Gophers and Tinygophers united if possible.

@soypat Hmmm, I think I'm missing something here.  I assumed TinyGo is downstream of Go with a direction arrow Go -> TinyGo.  I hadn't considered the opposite of TinyGo -> Go.  Are there examples in the past where TinyGo changes moved to upstream Go?  If the arrow can go both ways, netdev is not the right answer.

netdev would not make sense to integrate into upstream Go "net" package.  netdev is a hack.  It doesn't use the porting layer of Go's "net" package.  The porting layer is at the syscall interface to the OS.  We don't have an OS, so netdev slices above the porting layer, above the syscall interface.  The good news is applications written to Go's "net" package just work in TinyGo (sans features that haven't been ported).

BTW, I tried slipping in netdev at the porting layer to try to retain as much or all of "net" as possible.  I ran into two issues: first, it makes huge binaries.  I was able to compile some of the examples and was getting images that were ~10x bigger in flash and in RAM compared to what we have in this PR.  The full "net" package is huge, mind-numbingly huge (to quote DA).  The second issue was TLS.  netdev takes advantage of these co-processors to offload the TLS protocol to hardware/firmware.  That means the crypto/tls code in Go is pretty much by-passed in favor of direct netdev calls to hardware/firmware.  Go's porting layer doesn't comprehend TLS offload; it assumes TLS protocol is handled in software, and transported over TCP.

So here we are.  This PR provides a TinyGo "net" package that is a subset of Go's.  Just enough of the "net" package *.go files are copied over to give us the three basic net.Conns: TCPConn, UDPConn, TLSConn, as well as most of the "net/http" package.  It is a maintenance burden (keeping sync'ed with upstream Go "net").  But it's awesome also.  The examples for MQTT and websocket use the standard packages built on top of "net", and they just work.  But this PR is no good if the goal is to unify TinyGo and Go's "net" packages.

dgryski committed:
reflect: add SetMapIndex()

dgryski committed:
reflect: add MakeMap()

dgryski committed:
reflect: add MapKeys()

dgryski committed:
reflect: uncomment some DeepEqual tests that now pass

dgryski committed:
runtime: add sliceGrow function for reflect

scottfeldman commented:
One option is to move everything to drivers and have an empty "net" package in tinygo.  Apps that import "net" would now import "tinygo.org/x/drivers/net", same for "net/http".  (Kind of what's done now).  An app that imports just "net" will get an empty dir compile failure or something, and stick a src/net/README directing users to drivers/net.

That leaves tinygo's "net" namespace available in case something better comes along.

Would this help?

dgryski committed:
reflect: add Append()

dgryski committed:
reflect: add MakeSlice()

dgryski committed:
reflect: add Slice()

dgryski committed:
reflect: add Slice3()

dgryski committed:
reflect: add Bytes()

dgryski committed:
reflect: add Copy()

aykevl committed:
builder: remove non-ThinLTO build mode

All targets now support ThinLTO so let's remove the old unused code.

soypat commented:
~It's fine @scottfeldman you can put `netdev` under a internal directory for now (see [link](https://dave.cheney.net/2019/10/06/use-internal-packages-to-reduce-your-public-api-surface), same as above).~ Actually cant be done.

> I assumed TinyGo is downstream of Go with a direction arrow Go -> TinyGo. I hadn't considered the opposite of TinyGo -> Go. Are there examples in the past where TinyGo changes moved to upstream Go? If the arrow can go both ways, netdev is not the right answer.

It's not out of the question, as I believe TinyGo is gaining a lot of attention. That said it is highly unlikely changes would be merged upstream, especially this one which is a "hack".

> netdev would not make sense to integrate into upstream Go "net" package. netdev is a hack. It doesn't use the porting layer of Go's "net" package.

Agreed. But you are still including it in the form of a subpackage accesible to users, this is the same as if it were in upstream Go which can then break compatibility. We should avoid that by using the `internal` package functionality. See link above.

> BTW, I tried slipping in netdev at the porting layer to try to retain as much or all of "net" as possible.

Getting this working would be gold but I think it's a non-goal for this PR. We can strive for keeping the API identical with `net` package and Application usage identical that way we assume nothing about what the final solution to this `net` package dilemma will be (maybe using syscalls like BCG says or other solution).

> One option is to move everything to drivers and have an empty "net" package in tinygo. [...] That leaves tinygo's "net" namespace available in case something better comes along.

We should only do this if we can not maintain an **identical exported API** signature with upstream Go. As far as I can tell it is possible but requires the compiler `//go:linkname` directive and ~using the `internal` package.~ We should really make sure we are not slipping in exported API that is not in upstream Go IMHO so that we can move to a different solution in the future if something would require it.

soypat commented:
:facepalm: @scottfeldman   I just noticed if we put the `netdev` behind an `internal` package developers won't be able to develop. I have to think about this some more...

soypat commented:
I'll throw this possible solution I've come up with out there:


### Define the `socketer` interface in `net` but don't export it
This *sounds* like an idiomatic solution. 

1. Define `net.socketer` interface similarly to how you've done it but using standard library primitive types or existing `net` **exported** types. Do the same with `net.netdev` interface, which could be simply `net.dev` to avoid stutter.

2. Define `net.useDev` package function that takes a `net.dev` as an argument.

3. When one writes drivers one implements the `net.dev` interface which is doable since all of its types are primitive or `net` types. 

4. There could be a helper package called `netdev` somewhere outside of tinygo that helps users link their driver with the `net.useDev` package function using the `//go:linkname` compiler directive. i.e: `netdev.Use`. `netdev`  would have a interface mirroring the one in `net` (`net.dev`). 

### What this solves:
1. Tinygo <-> Go compatibility in full
2. Cyclic dependencies. Now there are none.
3. Developers can develop using external `netdev` package that mirrors `net.dev` interface

### Drawbacks
1. `net.socketer`/`netdev.Socketer`  interface is now composed of primitive types which impacts developer friendliness in a negative way.

aykevl committed:
compiler: add llvm.ident metadata

This metadata is emitted by Clang and I found it is important for source
level debugging on MacOS. This patch does not get source level debugging
to work yet (for that, it seems like packages need to be built
separately), but it is a step in the right direction.

deadprogram committed:
machine/stm32, nrf: flash API (#3472)

machine/stm32, nrf: implement machine.Flash

Implements the machine.Flash interface using the same definition as the tinyfs BlockDevice.

This implementation covers the stm32f4, stm32l4, stm32wlx, nrf51, nrf52, and nrf528xx processors.

deadprogram committed:
runtime/atsamd51: enable CMCC cache for greatly improved performance on SAMD51

Signed-off-by: deadprogram <ron@hybridgroup.com>

scottfeldman commented:
> I'll throw this possible solution I've come up with out there:
> 
> ### Define the `socketer` interface in `net` but don't export it
> This _sounds_ like an idiomatic solution.
> 
> 1. Define `net.socketer` interface similarly to how you've done it but using standard library primitive types or existing `net` **exported** types. Do the same with `net.netdev` interface, which could be simply `net.dev` to avoid stutter.
> 2. Define `net.useDev` package function that takes a `net.dev` as an argument.
> 3. When one writes drivers one implements the `net.dev` interface which is doable since all of its types are primitive or `net` types.

You lost me here.  How does a driver in the drivers repo get access to a non-exported interface in the tinygo repo?  The driver has to be written to some interface.

Maybe you can pull this PR (and the related driver one) and experiment with different options?  The success criteria would be that the examples in drivers/examples/net still compile and run, including the rawsocket example.

fgsch committed:
Fail earlier if Go is not available

dgryski committed:
reflect: add pointerTo()

dgryski committed:
reflect: allow nil rawType to call Kind()

dgryski committed:
reflect: add SetLen()

dgryski committed:
reflect: add Zero()

dgryski committed:
reflect: add Addr()

dgryski committed:
reflect: add OverflowFloat(), OverflowInt(), OverflowUint()

dgryski committed:
reflect: ensure all ValueError panics have Kind fields

dgryski committed:
compiler, reflect: add support for named types

dgryski committed:
reflect: add Type.String()

dgryski committed:
reflect: improve Value.String()

dgryski committed:
targets: bump cortex-m-qemu default stack size

dgryski committed:
reflect: add FieldByName(), and FieldByIndex()

dgryski committed:
reflect: add type check to Value.Field()

aykevl committed:
compiler: remove unsafe.Pointer(uintptr(v) + idx) optimization

I have checked this conversion is not needed anymore after the previous
commit, by running various smoke tests of which none triggered this
optimization. The only case where the optimization would have kicked in
is in syscall/syscall_windows.go:76 of the Go standard library.
Therefore, I prefer to remove it to reduce code complexity.

aykevl committed:
transform: fix non-determinism in the interface lowering pass

This non-determinism was introduced in
https://github.com/tinygo-org/tinygo/pull/2640. Non-determinism in the
compiler is a bug because it makes it harder to find whether a compiler
change actually affected the binary.

Fixes https://github.com/tinygo-org/tinygo/issues/3504

dgryski committed:
compiler, reflect: add Type.PkgPath

dgryski committed:
testdata: add brandondube/pctl to corpus

deadprogram committed:
machine/samd51: implement Flash interface

Signed-off-by: deadprogram <ron@hybridgroup.com>

dgryski committed:
os,syscall: Stat_t timespec fields are Atimespec on darwin

This allows archive/tar to build and pass.

dgryski committed:
syscall: add Timespec.Unix() for wasi.

This allows archive/tar to build (but not yet pass).

dgryski committed:
compiler, reflect: replace package and name length with null-byte termination

dgryski committed:
reflect: add support for remaining map types

paralin committed:
os: add IsTimeout function

Fixes build error: undefined: os.IsTimeout

Signed-off-by: Christian Stewart <christian@paral.in>

waj334 committed:
Refactor EnableInterrupts and DisableInterrupts

Removes usage of AsmFull which required an optimization pass to remove the map parameter passed into it. This caused issues when compiling with -opt=0 where memory for the map was being allocated as an unintended side-effect of using AsmFull with no optimizations enabled.

aykevl committed:
compiler: add debug info to []embed.files backing array

aykevl committed:
compiler: add debug information to //go:embed slice data

aykevl committed:
compiler: add debug information to []embed.file slice global

scottfeldman committed:
Use mirrored netdever interface

Changes in the commit based on feedback from Ben and Patricio:

0) Move netdever interface to "net" package
1) Mirror netdever interface with same in drivers repo
2) Use go:linkname to link to net.useNetdev()
3) Don't add any new Exports to "net" or "net/http" packages
4) Use native Go types in netdever interface
5) Add some common syscall types for things like AF_INET and SOCK_DGRAM.
6) Update documentation

scottfeldman committed:
Use mirrored netdever interface

Changes in the commit based on feedback from Ben and Patricio:

0) Move netdever interface to "net" package
1) Mirror netdever interface with same in drivers repo
2) Use go:linkname to link to net.useNetdev()
3) Don't add any new Exports to "net" or "net/http" packages
4) Use native Go types in netdever interface
5) Add some common syscall types for things like AF_INET and SOCK_DGRAM.
6) Update documentation

scottfeldman commented:
Build is busted because of F_SETFL symbol.  F_SETFL was added to src/syscall/net.go, as well as some other missing syscall constants, to support netdev interface using native Go types/constants.  Unfortunately, F_SETFL is defined in syscall_libc_darwin.go and syscall_libc_wasi.go, so not sure where the real F_SETFL should live.

Would someone please advise where to put a single F_SETFL to fix this build issue?

soypat commented:
Alright, will review in the coming weeks- 11k lines of code is quite a bunch

dgryski committed:
syscall: remove misleading comment about Stat_t fields

aykevl committed:
compiler; add position information to createConstant

Non-functional change. The position information will be used later to
emit debug info locations to string constants.

aykevl committed:
compiler: add debug location to string values

This is helpful for WebAssembly: it makes it possible to attribute many
more data to locations in the source code, which makes `-size=full` more
useful.

dgryski committed:
reflect: import visiblefields code and tests from upstream

aykevl committed:
compiler: try harder to find source locations for constants

If there is no source location, at least use the file (but not line) for
debug information.

This attributes almost 1kB of string data for ./testdata/stdlib.go when
compiling for WASI.

aykevl committed:
builder: detect compiler-rt size usage

Previously the code size of the compiler-rt library might be displayed
like this:

```
   1050       0       0       0 |    1050       0 | /home/ayke/src/tinygo/tinygo/llvm-project/compiler-rt/lib/builtins
    146       0       0       0 |     146       0 | /home/ayke/src/tinygo/tinygo/llvm-project/compiler-rt/lib/builtins/arm
```

With this change, it is displayed nicely like this:

```
   1196       0       0       0 |    1196       0 | C compiler-rt
```

deadprogram committed:
device/gba: add mostly complete hand-written register definitions

Signed-off-by: deadprogram <ron@hybridgroup.com>

deadprogram committed:
machine, runtime/interrupt: switch to use register definitions from device/gba

Signed-off-by: deadprogram <ron@hybridgroup.com>

aykevl committed:
builder: sizes: list interrupt vector as a pseudo-package for -size=full

This is another bit of memory that is now correctly accounted for in
`-size=full`.

aykevl committed:
builder: improve reading of DWARF variables

The compiler now uses DW_OP_plus_uconst in the address attribute of
DWARF variables. To be able to understand these addresses, we need to be
able to parse this opcode.

This commit should also improve the reliability of address parsing a
bit.

aykevl committed:
compiler: emit correct alignment in debug info for global variables

Previously we were using a really weird calculation to determine the
alignment in bits - written by me (no idea what I was thinking at the
time, it's obviously incorrect). Just replace it with the much more
obviously correct multiplication by 8 to get bits from bytes.

Found while working on properly dealing with alignment in `-size=full`.

aykevl committed:
compiler: account for alignment with -size=full

aykevl committed:
compiler: fix a race condition

There was a mostly benign race condition in the compiler. The issue was
that there is a check for type aliases (which can alias types in another
function), but this check was _after_ accessing a property of the
function that might not have been completed.

I don't think this can have any serious effects, as the function is
skipped anyway, but such bugs should certainly be fixed.

aykevl committed:
windows: disable parallelism on Windows

This is an attempt to figure out why the Windows CI keeps crashing. It
should be removed before the next release, by which point we should know
whether this has helped or not.

aykevl committed:
transform: add debug information to internal/task.stackSize

This has two benefits:
 1. It attributes these bytes to the internal/task package (in
    -size=full), instead of (unknown).
 2. It makes it possible to print the stack sizes variable in GDB.

This is what it might look like in GDB:

    (gdb) p 'internal/task.stackSizes'
    $13 = {344, 120, 80, 2048, 360, 112, 80, 120, 2048, 2048}

deadprogram committed:
build/linux: switch to ubuntu-20.04 for arm builds since ubuntu-18.04 is deprecated for GH actions runner

Signed-off-by: deadprogram <ron@hybridgroup.com>

dgryski committed:
reflect: disable visiblefields test for FieldByIndexErr

dgryski committed:
reflect: move StructField.Anonymous field to match upstream location

dgryski committed:
compiler, reflect: properly handle embedded structs

dgryski committed:
reflect: fix off-by-one in Zero sizing

Without this, pointers wouldn't be set to nil.  Add some tests.

dgryski committed:
reflect: set Index field in Field()

dgryski committed:
reflect: loosen unaddressable array rules for Copy()

sago35 committed:
machine/rp2040: correct issue with spi pin validation

sago35 committed:
machine/atsamd51: remove extra BK0RDY clear

dgryski committed:
reflect: MapIter.Next() needs to allocate new keys/values every time

dgryski committed:
compiler,reflect: add reflect.Type.NumMethods()

dgryski committed:
refect: Type.String() should use a shortened package name

deadprogram committed:
machine/samd21: implement Flash interface (#3496)

machine/samd21: implement Flash interface

dgryski committed:
reflect: add test for Type.NumMethod()

dgryski committed:
reflect: set PkgPath in StructField

dgryski committed:
compiler,reflect: fix pkgpath for struct fields

deadprogram committed:
machine/samd51: disable/restore Flash cache on write/erase

Signed-off-by: deadprogram <ron@hybridgroup.com>

scottfeldman committed:
get espat working with example tests

dgryski committed:
compiler: only define the package path once

Adding https://github.com/tinygo-org/tinygo/pull/3534 by hand to avoid conflicts when I rebase.

dgryski committed:
reflect: add stubs for Method(), CanConvert(), ArrayOf()

dgryski committed:
reflect: add UnsafePointer() for Func types

dgryski committed:
reflect: add UnsafeAddr()

dgryski committed:
reflect: add stub for StructOf()

dgryski committed:
debug: stub SetGCPercent()

deadprogram committed:
machine/flash: refactor to keep use of pure offset relative to start

Signed-off-by: deadprogram <ron@hybridgroup.com>

deadprogram committed:
machine/rp2040: correct Flash implementation

Signed-off-by: deadprogram <ron@hybridgroup.com>

deadprogram committed:
machine/stm32: correct Flash implementation

Signed-off-by: deadprogram <ron@hybridgroup.com>

sago35 committed:
machine/rp2040: change uart to allow for a single pin

sago35 committed:
wioterminal: fix pin definition of BCM13

dgryski committed:
reflect: Type.AssignableTo(): you can assign anything to interface{}

dgryski committed:
reflect: call decomposeInterface() directly in TypeOf()

dgryski committed:
reflect: make sure pointerTo() works for named types

dgryski committed:
reflect: fix Addr() indirect value/flags and add tests.

dgryski committed:
compiler,reflect: adjust struct layout for type info

dgryski committed:
compiler,reflect: sort out pkg path vs pkg name for named types

dgryski committed:
builder: fix binary size rodata value

dgryski committed:
reflect: TypeOf(nil) should be nil

dgryski committed:
transform: fix OptimizeReflectImplements pass for new named elem offset

dgryski committed:
reflect: add SetBytes()

dgryski committed:
Update src/reflect/value.go

Co-authored-by: Ayke <aykevanlaethem@gmail.com>

dgryski committed:
reflect: fix IsNil() for interfaces

dgryski committed:
reflect: fix Type.Name to return empty string for non-named types

    // Name returns the type's name within its package for a defined type.
    // For other (non-defined) types it returns the empty string.

aykevl committed:
transform: update reflect-implements test to opaque pointers

aykevl committed:
transform: update allocs test to opaque pointers

Also, rename most of the SSA values while we're at it.

aykevl committed:
transform: update gc-stackslots test to opaque pointers

aykevl committed:
transform: update interface test to opaque pointers

aykevl committed:
transform: update interrupt test to opaque pointers

aykevl committed:
transform: update maps test to opaque pointers

aykevl committed:
transform: update panic test to opaque pointers

aykevl committed:
transform: update stacksize test to opaque pointers

aykevl committed:
transform: update stringequal test to opaque pointers

aykevl committed:
transform: update stringtobytes test to opaque pointers

aykevl committed:
transform: update wasm-abi to use opaque pointers

dgryski committed:
builder: bump sizes tests

aykevl committed:
compiler: support all kinds of recursive types

Previously we only supported recursive types in structs. But there can
be other kinds of recursive types, like slices:

    type RecursiveSlice []RecursiveSlice

This doesn't involve structs, so it led to infinite recursion in the
compiler. This fix avoids recursion at the proper level: at the place
where the named type is defined.

dgryski committed:
reflect: stub MapOf()

dgryski committed:
reflect: fix isBinary() for float types

dgryski committed:
reflect: tweak Type.String() to match what encoding/json expects for empty structs

deadprogram committed:
docker: correct path for GHCR dev container build

Signed-off-by: deadprogram <ron@hybridgroup.com>

dgryski committed:
reflect: tweak Type.String() for interfaces to make encoding/xml happy

scottfeldman commented:
Hi @soypat, pester you on the review...how's it going?

soypat commented:
My apologies @scottfeldman, I totally missed the updated version. Thanks for letting me know. I'll get to it ASAP

scottfeldman committed:
pick up iprawsock additions

scottfeldman committed:
Merge branch 'netdev2-espat' into netdev2

soypat commented:
> Unfortunately, F_SETFL is defined in syscall_libc_darwin.go and syscall_libc_wasi.go, so not sure where the real F_SETFL should live.

Scott, the way the go tool works is that it interprets files that end with `_windows.go`, `_darwin.go`, `_linux.go`,  etc. as having an applied build tag corresponding to the GOOS environment variable. Similarly `_386.go` and `_amd64.go` suffixed files will have the GOARCH build tag applied. See https://pkg.go.dev/go/build#hdr-Build_Constraints

scottfeldman commented:
@soypat 

I have some updates in my local tree to this PR.  Should I push those or wait until your review?  Changes are:

- rebase with dev
- fix this F_SETFL issue
- get the espat driver working

soypat commented:
Push em!

scottfeldman committed:
define on one F_SETFL in syscall package

deadprogram commented:
Hello @scottfeldman looks like some actual CI errors like this https://github.com/tinygo-org/tinygo/actions/runs/4535422239/jobs/7990849307?pr=3452#step:20:34

scottfeldman commented:
I messed up my git repo for this PR, so I'm closing this PR and replacing it with a fresh one: https://github.com/tinygo-org/tinygo/pull/3614.
